### PR TITLE
add basic shell completion to `bin/scratch`

### PR DESCRIPTION
**Pull request by pH14**
I need all the help I can get with shell commands, so this adds in some shell autocompletion to `bin/scratch`. If it works well for everyone, we could take a similar approach for mzcompose and the other little CLIs we have.

Here's a little demo of it in action:

[asciicast](https://asciinema.org/a/VuMtCcwTivfGv4FFXgf2szB61)

TODO
~~add a section on how to generate and use the completion for bash and zsh into the readme~~ --> this will be added in a follow-up PR, that discusses all symlinkable changes https://github.com/MaterializeInc/materialize/pull/10609#issuecomment-1036649989
- [x] test it out on bash

### CODE REVIEW

**Review by benesch — COMMENTED** [review]
This is a wonderful idea! I would be strongly in favor of committing the generated completions for bash and zsh to `misc/completions/bash` and `misc/completions/zsh` and having CI verify that they stay up to date. Then folks who want completion can symlink that directory into their shell completions directory and have those completion files reflect the state of whatever commit they've got checked out at the moment.

File: misc/python/materialize/cli/scratch/__main__.py [hunk 1]
```diff
@@ -9,25 +9,51 @@
...
+            push.run,
+            "Push current HEAD (or a specific git commit) to scratch instance",
+        ),
     ]:
-        s = subparsers.add_parser(name)
+        s = subparsers.add_parser(name, description=description, help=description)
         configure(s)
         s.set_defaults(run=run)
 
+    shtab.add_argument_to(parser, ["-c", "--generate-completion"], parent=parser)  # type: ignore[no-untyped-call]
```
Comment (benesch) [inline misc/python/materialize/cli/scratch/__main__.py:55]: Super minor, but can we run this in subparser mode for consistency with the other commands?

```diff
diff --git a/misc/python/materialize/cli/scratch/__main__.py b/misc/python/materialize/cli/scratch/__main__.py
index c11e0f79f..b5e670f1c 100644
--- a/misc/python/materialize/cli/scratch/__main__.py
+++ b/misc/python/materialize/cli/scratch/__main__.py
@@ -47,13 +47,17 @@ def main() -> None:
             push.run,
             "Push current HEAD (or a specific git commit) to scratch instance",
         ),
+        (
+            "gen-completion",
+            lambda p: shtab.add_argument_to(p, "generate-completion", parent=parser),
+            lambda: None,
+            "Generate shell completion script",
+        ),
     ]:
         s = subparsers.add_parser(name, description=description, help=description)
         configure(s)
         s.set_defaults(run=run)
 
-    shtab.add_argument_to(parser, ["-c", "--generate-completion"], parent=parser)  # type: ignore[no-untyped-call]
-
     args = parser.parse_args()
     args.run(args)
 
```

Comment (benesch) [reply misc/python/materialize/cli/scratch/__main__.py:55]: Oh, also, I try very hard to stay away from `# type: ignore` comments! Here's a stub file you can add instead:

```diff
diff --git a/misc/python/stubs/shtab.pyi b/misc/python/stubs/shtab.pyi
new file mode 100644
index 000000000..7970f48d0
--- /dev/null
+++ b/misc/python/stubs/shtab.pyi
@@ -0,0 +1,10 @@
+from argparse import ArgumentParser
+from typing import List, Optional, Union
+
+def add_argument_to(
+    parser: ArgumentParser,
+    option_string: Union[str, List[str]] = ...,
+    help: str = ...,
+    parent: Optional[ArgumentParser] = ...,
+    preamble: str = ...,
+) -> None: ...

```

File: bin/completion [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+              exit 1
+            fi
+          done
+        done
+        ;;
+    *)
+        printf "unknown command %q\n" "$cmd"
+        exit 1
+        ;;
+esac
```
Comment (benesch) [inline bin/completion:71]: This is a very well written shell script, and it's short enough to probably stay in shell forever, but heads up that I've been increasingly reaching towards writing these scripts in Python from the get go (e.g. https://github.com/MaterializeInc/materialize/blob/e49579ca8dd8e227e13bfd58df707d1ccf0e07b3/bin/ci-builder#L14-L15).

**Review by benesch — APPROVED** [review]
So good!

File: misc/python/materialize/cli/scratch/__main__.py (hunk 1 above)
Comment (pH14) [reply misc/python/materialize/cli/scratch/__main__.py:55]: Oooooh that's sweet, thanks for the tip. everything about the python >2.7 ecosystem is new and shiny to me 😁

File: bin/completion (hunk 2 above)
Comment (pH14) [reply bin/completion:71]: 👍 I might leave this for now, but will add a readme to `bin` that recommends the use of python for anything beyond a trivial shell script

**Pull request merged**